### ISSUE TITLE
[Datasets] Deprecate `flat_map`

### ISSUE THREAD

**Issue by bveeramani** [ray-project/ray#30393]
### Description

Deprecate and then eventually remove `ray.data.Dataset.flat_map`.

### Use case

There're many map-related methods:
* `map`
* `map_batches`
* `flat_map`

It's confusing. 

We can make dataset transformations less confusing by eliminating `flat_map`. Few people use `flat_map`, so removing it shouldn't be an issue.

**Comment by ericl** [comment]
We can't remove flat_map, as it's a fundamental dataset operation. We would also remove map at the same time.

**Comment by bveeramani** [comment]
@ericl I had discussed this with @clarkzinzow before opening an issue. 

Could you elaborate on why we need to remove `flat_map` and `map` at the same time?

**Comment by ericl** [comment]
To elaborate here:
1) flat_map is a fundamental dataset operation. A large number of workloads are not possible to express without using flat_map, in addition to map.
2) Users coming from other data systems like tf.Data / Spark will expect this functionality.
3) It's technically possible to also express this in map_batches(), but the same is true for map()--- so this is an inconsistent decision.

In short, flat_map() is an important feature that we cannot "just remove" because some users haven't heard of it before.

**Comment by Wendi-anyscale** [comment]
This is a high-severity issue from the ux assessment. @ericl do you think we would have any good ways to solve the confusion at the API level? Otherwise, Updating the doc with a comparison table of these 3 options, ideally clear decision-making tree designed for the users to choose which one to use would be a temporary fix.

**Comment by bveeramani** [comment]
Closing per offline discussion
